core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java

```diff
@@ -161,33 +154,38 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         }
         String groupBy = params.param("group_by", "nodes");
         if ("nodes".equals(groupBy)) {
+            if (discoveryNodes == null) {
+                throw new IllegalStateException("discoveryNodes must be set before calling toXContent with group_by=nodes");
+            }
             builder.startObject("nodes");
-            for (Map.Entry<DiscoveryNode, List<TaskInfo>> entry : getPerNodeTasks().entrySet()) {
-                DiscoveryNode node = entry.getKey();
-                builder.startObject(node.getId());
-                builder.field("name", node.getName());
-                builder.field("transport_address", node.getAddress().toString());
-                builder.field("host", node.getHostName());
-                builder.field("ip", node.getAddress());
-
-                builder.startArray("roles");
-                for (DiscoveryNode.Role role : node.getRoles()) {
-                    builder.value(role.getRoleName());
-                }
-                builder.endArray();
-
-                if (!node.getAttributes().isEmpty()) {
-                    builder.startObject("attributes");
-                    for (Map.Entry<String, String> attrEntry : node.getAttributes().entrySet()) {
-                        builder.field(attrEntry.getKey(), attrEntry.getValue());
+            for (Map.Entry<String, List<TaskInfo>> entry : getPerNodeTasks().entrySet()) {
+                DiscoveryNode node = discoveryNodes.get(entry.getKey());
+                builder.startObject(entry.getKey());
+                if (node != null) {
+                    // If the node is no longer part of the cluster, oh well, we'll just skip it's useful information.
```

Comment:
I don't feel too bad about it because it should be super infrequent. I did it to make working with the TaskInfo in all the cases where you don't need the node's full information easier. And to shrink the wire format, but that wasn't that important to me.